src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/DesignTimeBasicWriter.cs

```diff
@@ -153,9 +154,110 @@ public override void WriteHtmlAttributeValue(CSharpRenderingContext context, Htm
             context.RenderChildren(node);
         }
 
-        public override void WriteCSharpAttributeValue(CSharpRenderingContext context, CSharpAttributeValueIRNode node)
+        public override void WriteCSharpExpressionAttributeValue(CSharpRenderingContext context, CSharpExpressionAttributeValueIRNode node)
         {
-            context.RenderChildren(node);
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            if (node.Children.Count == 0)
+            {
+                return;
+            }
+
+            var firstChild = node.Children[0];
+            if (firstChild.Source != null)
+            {
+                using (context.Writer.BuildLinePragma(firstChild.Source.Value))
+                {
+                    var offset = RazorDesignTimeIRPass.DesignTimeVariable.Length + " = ".Length;
+                    context.Writer.WritePadding(offset, firstChild.Source, context);
+                    context.Writer.WriteStartAssignment(RazorDesignTimeIRPass.DesignTimeVariable);
+
+                    for (var i = 0; i < node.Children.Count; i++)
+                    {
+                        if (node.Children[i] is RazorIRToken token && token.IsCSharp)
+                        {
+                            context.AddLineMappingFor(token);
+                            context.Writer.Write(token.Content);
+                        }
+                        else
+                        {
+                            // There may be something else inside the expression like a Template or another extension node.
+                            context.RenderNode(node.Children[i]);
+                        }
+                    }
+
+                    context.Writer.WriteLine(";");
+                }
+            }
+            else
+            {
+                context.Writer.WriteStartAssignment(RazorDesignTimeIRPass.DesignTimeVariable);
+                for (var i = 0; i < node.Children.Count; i++)
+                {
+                    if (node.Children[i] is RazorIRToken token && token.IsCSharp)
+                    {
+                        context.Writer.Write(token.Content);
```

Comment:
Still need to conditionally add a line mapping for each token. Otherwise, any tokens that have line mappings wont get design time intellisense.